### DISCUSSION TITLE
Usage do AffineTransform

### DISCUSSION THREAD

**Discussion by a-parida12** [Project-MONAI/MONAI#5241] [Q&A]
I use the `SpacingD` transforms after  `LoadImageD`  and `AddChannelD`. I read in a nifti file of shape [368, 336, 368] , the input to the `spacingD` [1, 368, 336, 368]. The outputs of spacingD is a matrix with zeros.

I followed the trace and the last time I see non-zeros matrix is at 

https://github.com/Project-MONAI/MONAI/blob/dev/monai/networks/layers/spatial_transforms.py#L575

The generated grid has a range [1,3] but the `torch.grid_sample` expects range [-1, 1].
More details here- 
https://pytorch.org/docs/stable/generated/torch.nn.functional.affine_grid.html
https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html

I tried it various nifti files and gives me the same trouble. When I change the range to the expected range by substracting 2. i get a output, which looks good. 

Can anyone verify the issue. I am using MonAi 1.0.0

**Comment by wyli** [comment]
if you share one of the nifti files or the header info, I can look into this.

**Reply by a-parida12** [reply]
I will put the nibabel header dump here.

**Reply by a-parida12** [reply]
<img width="826" alt="Screenshot 2022-10-01 at 11 36 48" src="https://user-images.githubusercontent.com/23334003/193416978-b5c06ecb-d273-44ca-8b93-9d4afd412f9f.png">

**Reply by wyli** [reply]
thanks, the `pixdim` field looks suspicious, for spatial 3d image, pixdim[0] should be 3, but it's 1:
![Screenshot 2022-10-03 at 09 05 03](https://user-images.githubusercontent.com/831580/193529218-7cd758de-a560-4a67-8e2c-2ca6bbfd4d27.png)

**Reply by a-parida12** [reply]
I am not entirely sure. I think  has to be `dim[0] == 3` and not `pixdim[0]`. Also here is an example header from the internet https://dartbrains.org/content/Introduction_to_Neuroimaging_Data.html#loading-data-with-nibabel .

**Reply by wyli** [reply]
right, the header should be fine, then the isuse might be `AddChanneld` which is deprecated. Have you tried `EnsureChannelFirstD`? the following example works fine:

```py
import numpy as np
import nibabel as nib
import monai.transforms as mt

img = np.ones((368, 336, 368)).astype(np.float32)
a = np.asarray([[0.41, 0.005, 0.008, -79.7], [-0.0049, 0.592, 0.0664, -57.4], [-0.0073, -0.0972, 0.404, -32.1], [0., 0., 0., 1.0]])
nib_img = nib.Nifti1Image(img, affine=a)
nib.save(nib_img, "test.nii.gz")
print(nib_img)

x = mt.LoadImageD("image")({"image": "test.nii.gz"})
x = mt.EnsureChannelFirstD("image")(x)
x = mt.SpacingD("image", (1.0, 1.0, 1.0))(x)
print(x["image"])
print(x["image"].shape)
print(x["image"].pixdim)
```

**Reply by a-parida12** [reply]
For me both the `EnsureChannelFirstD` and the `AddChannelD` give the same output. Here is a small snippet below-

```python
import monai.transforms as mt
import numpy as np

x = mt.LoadImageD("image")({"image": "/home/abhijeet/Datasets/123.nii.gz"})
x = mt.EnsureChannelFirstD("image")(x)
print(np.min(x["image"]), np.max(x["image"]))
# 0.0 2740.2239
x = mt.SpacingD("image", (1.0, 1.0, 1.0))(x)
print(np.min(x["image"]), np.max(x["image"]))
# 0.0 4.2632482e-06

```
I would be happy to send the file to you if it helps. Let me know I can send it to you it is 160MB file.

**Reply by wyli** [reply]
thanks, I looked into this, probably it's a bug from computing the offset, I'll submit a PR to revise it.

**Reply by a-parida12** [reply]
Thank you @wyli !

**Comment by a-parida12** [comment]
@wyli I have similar troubles of the matrix becoming zeros after a rotation operation using `RandRotateD` is it also related to the issue you looked into before with the offset. 

Also do you have a release or a nightly build that I could use with the changes

**Reply by wyli** [reply]
sure, if you are using the docker image the `latest` tag has the patch https://hub.docker.com/r/projectmonai/monai/tags. otherwise you can pip install with the `git+` syntax https://docs.monai.io/en/latest/installation.html#from-github

**Reply by a-parida12** [reply]
Thanks again!!
